Textile/Classes/RequestsHandler.m

```diff
@@ -0,0 +1,366 @@
+//
+//  PinManager.m
+//  Textile
+//
+//  Created by Aaron Sutula on 5/3/19.
+//
+
+#import "RequestsHandler.h"
+#import "TextileApi.h"
+#import "Callback.h"
+
+@interface RequestsHandler () <NSURLSessionDelegate, NSURLSessionTaskDelegate, NSURLSessionDataDelegate>
+
+@property (nonatomic, strong) NSURLSession *session;
+
+@end
+
+@implementation RequestsHandler
+
+dispatch_queue_t flushQueue;
+
+- (instancetype)init {
+  if (self = [super init]) {
+    // This serial queue is used to process calls to flush so we only process one at a time
+    flushQueue = dispatch_queue_create("io.textile.flushQueue", DISPATCH_QUEUE_SERIAL);
+
+    // Configure and create our NSURLSession used for uplaods
+    NSURLSessionConfiguration *config = [NSURLSessionConfiguration backgroundSessionConfigurationWithIdentifier:TEXTILE_BACKGROUND_SESSION_ID];
+    config.networkServiceType = NSURLNetworkServiceTypeResponsiveData;
+    config.allowsCellularAccess = YES; // default
+    config.timeoutIntervalForRequest = 60; //default
+    config.timeoutIntervalForResource = 60 * 60 * 24; // 1 day
+    // config.waitsForConnectivity = YES; // defaults to YES for background sessions, not availiable until iOS 11 anyway
+    config.sessionSendsLaunchEvents = YES; // default
+    config.discretionary = NO; // default
+    config.shouldUseExtendedBackgroundIdleMode = YES; // not really sure about this
+    config.HTTPMaximumConnectionsPerHost = 10; // default 4 for ios
+    self.session = [NSURLSession sessionWithConfiguration:config delegate:self delegateQueue:nil];
+  }
+  return self;
+}
+
+- (void)flush {
+  // We don't know what thread we're being called on here, so dispatch to our
+  // serial queue to make sure only one call to flush can be processed at a time
+  dispatch_async(flushQueue, ^{
+    [self processQueue];
+  });
+}
+
+- (void)processQueue {
+  NSLog(@"Flushing");
+
+  NSError *error;
+  NSData *cafeRequestsData = [self.node cafeRequests:-1 error:&error];
+  if (error) {
+    NSLog(@"cafeRequests error: %@", error.localizedDescription);
+    return;
+  }
+
+  Strings *requestIds = [[Strings alloc] initWithData:cafeRequestsData error:&error];
+  if (!requestIds) {
+    NSLog(@"Error deserializing Strings: %@", error.localizedDescription);
+    return;
+  }
+
+  // Dispatch group used to wait until processing all CafeHTTPRequests in this call to process are complete
+  dispatch_group_t group = dispatch_group_create();
+
+  for (NSString *requestId in requestIds.valuesArray) {
+
+    // TODO: Is there any reason to call failCafeRequest before it is marked as pending?
+    // TODO: Should we mark as pending before writing to disk or after?
```

Comment:
Would like to have answers to these questions just to be sure. My current implementation doesn't call `failCafeRequest` if the request has not already been marked as pending and marking as pending happens after writing the request to disk, but before the upload starts.

Reply:
Covered in call... in Java land I opted to mark as pending ASAP, then on error, mark as not pending.